tests/operators/test_offres_exp_approx.py

```diff
@@ -0,0 +1,241 @@
+"""Test off-resonance spatial coefficient and temporal interpolator estimation."""
+
+import math
+
+import numpy as np
+import numpy.testing as npt
+
+import pytest
+
+import mrinufft
+from mrinufft._utils import get_array_module
+
+CUPY_AVAILABLE = True
+try:
+    import cupy as cp
+except ImportError:
+    CUPY_AVAILABLE = False
+
+TORCH_AVAILABLE = True
+try:
+    import torch
+except ImportError:
+    TORCH_AVAILABLE = False
+
+# parameter combinations (ndims, backend, device)
+params = [(2, "cpu", None), (3, "cpu", None)]
+
+if CUPY_AVAILABLE:
+    params.extend([(2, "gpu", None), (3, "gpu", None)])
+
+if TORCH_AVAILABLE:
+    params.extend([(2, "torch", "cpu"), (3, "torch", "cpu")])
+
+if TORCH_AVAILABLE and CUPY_AVAILABLE:
+    params.extend([(2, "torch", "cuda"), (3, "torch", "cuda")])
+
+
+def calculate_true_offresonance_term(fieldmap, t):
+    """Calculate non-approximate off-resonance modulation term."""
+    xp = get_array_module(fieldmap)
+    t = xp.asarray(t)
+    if xp.__name__ == "torch":
+        t = t.to(device=fieldmap.device)
+    arg = t * fieldmap[..., None]
+    arg = arg[None, ...].swapaxes(0, -1)[..., 0]
+    return xp.exp(-arg)
+
+
+def calculate_approx_offresonance_term(B, C):
+    """Calculate approximate off-resonance modulation term."""
+    field_term = 0.0
+    for n in range(B.shape[0]):
+        tmp = B[n] * C[n][..., None]
+        tmp = tmp[None, ...].swapaxes(0, -1)[..., 0]
+        field_term += tmp
+
+    return field_term
+
+
+def make_b0map(ndim, npix=64):
+    """Make B0 map (units: Hz)."""
+    # generate support
+    shape = ndim * [npix]
+    if ndim == 2:
+        mask, fieldmap = _make_disk(shape)
+    elif ndim == 3:
+        mask, fieldmap = _make_sphere(shape)
+
```

Comment:
I like these functions, it could help to have it somewhere internally supported. Also @paquiteau it could help to add your complex B0 map maker from brainweb-dl somewhere (maybe it belongs in brainweb-dl, not sure..)

Reply:
I don't have support (yet) for B0 map in `brainweb-dl` (but 3 maps are available for download https://brainweb.bic.mni.mcgill.ca/about_sbd.html). Generating trivial B0-maps could be setup in `mrinufft.extras` (along with some smaps generation code I keep copy pasting everywhere)